### ISSUE TITLE
terragrunt-cache behavior change for nested dependencies (since v0.55.6)

### ISSUE THREAD

**Issue by ubbeK** [gruntwork-io/terragrunt#3028]
**Describe the bug**
Since terragrunt version [0.55.6](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.55.6), a change of terragrunt-cache behavior has been introduced when it comes to multi level dependencies. If we have 3 different terragrunt root modules A, B and C, where B is dependent on A, and C is dependent on B (i.e. `C -> B -> A`), then the `terragrunt init` on C will try to cache module A and fail while doing so. This was not the case prior to `0.55.6`.

**To Reproduce**

Files hierarchy and content

```
urban@ubbeK:~/terragrunt $ tree .
.
├── bin
│   ├── terragrunt-0.55.5
│   └── terragrunt-0.55.6
├── modules
│   ├── roots
│   │   ├── a
│   │   │   └── terragrunt.hcl
│   │   ├── b
│   │   │   └── terragrunt.hcl
│   │   └── c
│   │       └── terragrunt.hcl
│   └── submodules
│       ├── a
│       │   └── main.tf
│       ├── b
│       │   └── main.tf
│       └── c
└──         └── main.tf
```

```hcl
# ./modules/roots/a/terragrunt.hcl
terraform {
  source = "${get_terragrunt_dir()}/../../submodules/a"
}

# ./modules/roots/b/terragrunt.hcl
terraform {
  source = "${get_terragrunt_dir()}/../../submodules/b"
}

dependency "tf" {
  config_path = "${get_terragrunt_dir()}/../a"
}

# ./modules/roots/c/terragrunt.hcl
terraform {
  source = "${get_terragrunt_dir()}/../../submodules/c"
}

dependency "tf" {
  config_path = "${get_terragrunt_dir()}/../b"
}

# ./modules/submodules/a/main.tf
output "family_name" {
  value = "a"
}

# ./modules/submodules/b/main.tf
output "family_name" {
  value = "b"
}

# ./modules/submodules/c/main.tf
output "family_name" {
  value = "c"
}
```

Commands executed on version `0.55.6`

```bash
~/terragrunt/bin/terragrunt-0.55.6 --version

cd ~/terragrunt
rm -rf ~/terragrunt/modules/roots/a/.terragrunt-cache
rm -rf ~/terragrunt/modules/roots/b/.terragrunt-cache
rm -rf ~/terragrunt/modules/roots/c/.terragrunt-cache

cd ~/terragrunt/modules/roots/a
~/terragrunt/bin/terragrunt-0.55.6 init
~/terragrunt/bin/terragrunt-0.55.6 apply -auto-approve

cd ~/terragrunt/modules/roots/b
~/terragrunt/bin/terragrunt-0.55.6 init
~/terragrunt/bin/terragrunt-0.55.6 apply -auto-approve

cd ~/terragrunt/modules/roots/c
~/terragrunt/bin/terragrunt-0.55.6 init
```

**Expected behavior**

`terragrunt init` should not fail on module `C`.

**Nice to have**
- [ ] Terminal output
```
urban@ubbeK:~/terragrunt $ ~/terragrunt/bin/terragrunt-0.55.6 --version
terragrunt version v0.55.6
urban@ubbeK:~/terragrunt $ cd ~/terragrunt
urban@ubbeK:~/terragrunt $ rm -rf ~/terragrunt/modules/roots/a/.terragrunt-cache
urban@ubbeK:~/terragrunt $ rm -rf ~/terragrunt/modules/roots/b/.terragrunt-cache
urban@ubbeK:~/terragrunt $ rm -rf ~/terragrunt/modules/roots/c/.terragrunt-cache
urban@ubbeK:~/terragrunt $ cd ~/terragrunt/modules/roots/a
urban@ubbeK:~/terragrunt/modules/roots/a $ ~/terragrunt/bin/terragrunt-0.55.6 init
INFO[0000] Downloading Terraform configurations from file:///home/urban/terragrunt/modules/submodules/a into /home/urban/terragrunt/modules/roots/a/.terragrunt-cache/z_GCSxwnc8oez_ZWtcQAd8jUNAQ/FEiijTEI50xqruhwi8IhNp90MwY 

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
urban@ubbeK:~/terragrunt/modules/roots/a $ ~/terragrunt/bin/terragrunt-0.55.6 apply -auto-approve

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

Changes to Outputs:
  + family_name = "a"

You can apply this plan to save these new output values to the Terraform
state, without changing any real infrastructure.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

family_name = "a"
urban@ubbeK:~/terragrunt/modules/roots/a $ cd ~/terragrunt/modules/roots/b
urban@ubbeK:~/terragrunt/modules/roots/b $ ~/terragrunt/bin/terragrunt-0.55.6 init
INFO[0000] Downloading Terraform configurations from file:///home/urban/terragrunt/modules/submodules/b into /home/urban/terragrunt/modules/roots/b/.terragrunt-cache/5CF2RMIa1wfIA10OGn0SYtjkacg/1RpYqPm8FsqkUGpchfcP7cMzIsI 

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
urban@ubbeK:~/terragrunt/modules/roots/b $ ~/terragrunt/bin/terragrunt-0.55.6 apply -auto-approve

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

Changes to Outputs:
  + family_name = "b"

You can apply this plan to save these new output values to the Terraform
state, without changing any real infrastructure.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

family_name = "b"
urban@ubbeK:~/terragrunt/modules/roots/b $ cd ~/terragrunt/modules/roots/c
urban@ubbeK:~/terragrunt/modules/roots/c $ ~/terragrunt/bin/terragrunt-0.55.6 init
INFO[0000] Downloading Terraform configurations from file:///home/urban/terragrunt/modules/submodules/a into /home/urban/terragrunt/modules/roots/c/.terragrunt-cache/z_GCSxwnc8oez_ZWtcQAd8jUNAQ/FEiijTEI50xqruhwi8IhNp90MwY  prefix=[/home/urban/terragrunt/modules/roots/a] 
WARN[0000] Error reading partial config  for dependency tf: /home/urban/terragrunt/modules/roots/a/terragrunt.hcl is a dependency of /home/urban/terragrunt/modules/roots/b/terragrunt.hcl but detected no outputs. Either the target module has not been applied yet, or the module has no outputs. If this is expected, set the skip_outputs flag to true on the dependency block. 
ERRO[0000] /home/urban/terragrunt/modules/roots/a/terragrunt.hcl is a dependency of /home/urban/terragrunt/modules/roots/b/terragrunt.hcl but detected no outputs. Either the target module has not been applied yet, or the module has no outputs. If this is expected, set the skip_outputs flag to true on the dependency block. 
ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```
- [ ] Screenshots
![image](https://github.com/gruntwork-io/terragrunt/assets/8124023/7b9ddb63-6d1a-49dc-b61c-5283e9a24efc)

**Versions**
- Terragrunt version: 0.55.6
- Terraform version: v1.7.5
- Environment details (Ubuntu 20.04, Windows 10, etc.): `Ubuntu 22.04.4 LTS` as WSL on Windows 10

**Additional context**

I have done the same test using terragrunt version `0.55.5` with good results, meaning that the following PR is probably whats causing it: https://github.com/gruntwork-io/terragrunt/pull/2957

here are the succeeding commands and console output for terragrunt `0.55.5` using the same files

```bash
~/terragrunt/bin/terragrunt-0.55.5 --version

cd ~/terragrunt
rm -rf ~/terragrunt/modules/roots/a/.terragrunt-cache
rm -rf ~/terragrunt/modules/roots/b/.terragrunt-cache
rm -rf ~/terragrunt/modules/roots/c/.terragrunt-cache

cd ~/terragrunt/modules/roots/a
~/terragrunt/bin/terragrunt-0.55.5 init
~/terragrunt/bin/terragrunt-0.55.5 apply -auto-approve

cd ~/terragrunt/modules/roots/b
~/terragrunt/bin/terragrunt-0.55.5 init
~/terragrunt/bin/terragrunt-0.55.5 apply -auto-approve

cd ~/terragrunt/modules/roots/c
~/terragrunt/bin/terragrunt-0.55.5 init
~/terragrunt/bin/terragrunt-0.55.5 apply -auto-approve
```

```
urban@ubbeK:~/terragrunt $ ~/terragrunt/bin/terragrunt-0.55.5 --version
terragrunt version v0.55.5
urban@ubbeK:~/terragrunt $ cd ~/terragrunt
urban@ubbeK:~/terragrunt $ rm -rf ~/terragrunt/modules/roots/a/.terragrunt-cache
urban@ubbeK:~/terragrunt $ rm -rf ~/terragrunt/modules/roots/b/.terragrunt-cache
urban@ubbeK:~/terragrunt $ rm -rf ~/terragrunt/modules/roots/c/.terragrunt-cache
urban@ubbeK:~/terragrunt $ cd ~/terragrunt/modules/roots/a
urban@ubbeK:~/terragrunt/modules/roots/a $ ~/terragrunt/bin/terragrunt-0.55.5 init
INFO[0000] Downloading Terraform configurations from file:///home/urban/terragrunt/modules/submodules/a into /home/urban/terragrunt/modules/roots/a/.terragrunt-cache/z_GCSxwnc8oez_ZWtcQAd8jUNAQ/FEiijTEI50xqruhwi8IhNp90MwY 

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
urban@ubbeK:~/terragrunt/modules/roots/a $ ~/terragrunt/bin/terragrunt-0.55.5 apply -auto-approve

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

Changes to Outputs:
  + family_name = "a"

You can apply this plan to save these new output values to the Terraform
state, without changing any real infrastructure.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

family_name = "a"
urban@ubbeK:~/terragrunt/modules/roots/a $ cd ~/terragrunt/modules/roots/b
urban@ubbeK:~/terragrunt/modules/roots/b $ ~/terragrunt/bin/terragrunt-0.55.5 init
INFO[0000] Downloading Terraform configurations from file:///home/urban/terragrunt/modules/submodules/b into /home/urban/terragrunt/modules/roots/b/.terragrunt-cache/5CF2RMIa1wfIA10OGn0SYtjkacg/1RpYqPm8FsqkUGpchfcP7cMzIsI 

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
urban@ubbeK:~/terragrunt/modules/roots/b $ ~/terragrunt/bin/terragrunt-0.55.5 apply -auto-approve

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

Changes to Outputs:
  + family_name = "b"

You can apply this plan to save these new output values to the Terraform
state, without changing any real infrastructure.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

family_name = "b"
urban@ubbeK:~/terragrunt/modules/roots/b $ cd ~/terragrunt/modules/roots/c
urban@ubbeK:~/terragrunt/modules/roots/c $ ~/terragrunt/bin/terragrunt-0.55.5 init
INFO[0000] Downloading Terraform configurations from file:///home/urban/terragrunt/modules/submodules/c into /home/urban/terragrunt/modules/roots/c/.terragrunt-cache/_xQH-j39DA7C0Xdt0DCDf4mdL30/p93U5EiqFzs6wR6HAwapQQEi2sA 

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
urban@ubbeK:~/terragrunt/modules/roots/c $ ~/terragrunt/bin/terragrunt-0.55.5 apply -auto-approve

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

Changes to Outputs:
  + family_name = "c"

You can apply this plan to save these new output values to the Terraform
state, without changing any real infrastructure.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

family_name = "c"
```

**Comment by thrix** [comment]
Yep, it is broken, I am hitting it during run-all destroy with the latest 0.57.2 version. Reverting back to 0.55.5 fixes the issuefor me. I have a transitive dependency `C -> B -> A` and it is picking up a mock output from `A` instead of the real output.

**Comment by levkohimins** [comment]
@ubbeK, thanks for the detail explanation!

**Comment by levkohimins** [comment]
Hi @bgrommes, resolved in [v0.58.10](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.58.10) release.
